tests/test_insert_deletes_consolidate.cpp

```diff
@@ -79,59 +75,115 @@ inline void load_aligned_bin_part(const std::string& bin_file, T* data,
 }
 
 std::string get_save_filename(const std::string& save_path,
-                              size_t points_to_skip, size_t points_deleted,
+                              size_t points_to_skip, float points_deleted,
+                              float  points_inserted,
                               size_t last_point_threshold) {
   std::string final_path = save_path;
   if (points_to_skip > 0) {
     final_path += "skip" + std::to_string(points_to_skip) + "-";
   }
-  if (points_deleted > 0) {
-    final_path += "del" + std::to_string(points_deleted) + "-";
+  if (points_deleted > 0 && points_inserted > 0) {
+    final_path += "insert" + std::to_string(points_inserted) + "-" + "delete" +
+                  std::to_string(points_deleted);
+  } else if (points_deleted == 0 && points_inserted > 0) {
+    final_path += "insert" + std::to_string(points_inserted);
+  } else if (points_deleted > 0 && points_inserted == 0) {
+    final_path += "delete" + std::to_string(points_deleted);
   }
-  final_path += std::to_string(last_point_threshold);
   return final_path;
 }
 
 template<typename T, typename TagT>
-void insert_till_next_checkpoint(diskann::Index<T, TagT>& index, size_t start,
-                                 size_t end, size_t thread_count, T* data,
-                                 size_t aligned_dim) {
+void insert_till_next_checkpoint(diskann::Index<T, TagT>& index,
+                                 size_t max_points, float insert_percentage,
+                                 float build_percentage, size_t thread_count,
+                                 T* data, size_t aligned_dim,
+                                 float                delete_percentage,
+                                 diskann::Parameters& delete_params) {
   diskann::Timer insert_timer;
+  int64_t        insert_points = max_points * insert_percentage;
+  int64_t        delete_points = max_points * delete_percentage;
+  int64_t        build_points = max_points * build_percentage;
+  if (delete_percentage == insert_percentage) {
+    std::cout << " random insertion and deletion" << std::endl;
+    std::vector<int> streaming_numbers;
+    for (int64_t j = 0; j < (int64_t) build_points - 1; j++) {
+      streaming_numbers.push_back(j);
+    }
+    random_shuffle(streaming_numbers.begin(), streaming_numbers.end());
 
 #pragma omp parallel for num_threads(thread_count) schedule(dynamic)
-  for (int64_t j = start; j < (int64_t) end; j++) {
-    index.insert_point(&data[(j - start) * aligned_dim], static_cast<TagT>(j));
+    for (int64_t i1 = 0; i1 < insert_points; i1++) {
+      int64_t k1 = streaming_numbers[i1];
+      index.insert_point(&data[k1 * aligned_dim], static_cast<TagT>(k1));
```

Comment:
It's not random enough, you can only insert a subset data beginning with 0, random_shuffle only can change the insert order.

Reply:
My random insertion scheme is that when there are insert and delete, it can be randomly selected from the complete set. For only insert, since the full amount needs to be kept with the build, if I randomly take the data set during build, then the data set when I insert is equivalent to random (for example, the full amount is 100,000, and the front of the random data set is taken during build. 70,000, and insert takes the last 30,000 data sets)